src/app/templates/InputGeneral/SelectGeneral.tsx

```diff
@@ -0,0 +1,219 @@
+          setOpened={setOpened}
+          {...optionsProps}
+        />
+      )}
+    >
+      {({ ref, opened, toggleOpened }) => (
+        <div ref={ref as unknown as React.RefObject<HTMLDivElement>}>
+          {opened ? (
+            <SelectSearch className={dropdownButtonClassName} {...searchProps} />
+          ) : (
+            <div
+              className="w-full flex items-center justify-between border rounded-md border-gray-300"
+              style={{ maxHeight: '4.5rem', overflow: 'hidden' }}
+            >
+              <button
+                className={classNames(
+                  'flex gap-2 items-center',
+                  isInputDefined && 'border-r border-gray-300',
+                  !isInputDefined && 'w-full justify-between',
+                  dropdownButtonClassName
+                )}
+                onClick={() => {
+                  toggleOpened();
+                  trackDropdownClick();
+                }}
+                {...setTestID(testIds?.dropdownTestId)}
+              >
+                {DropdownFaceContent}
+                <ChevronDownIcon className="text-gray-600 stroke-current stroke-2" style={{ height: 16, width: 16 }} />
+              </button>
+              {Input}
+            </div>
+          )}
+        </div>
+      )}
+    </Popper>
+  );
+};
+interface SelectOptionsPropsBase<Type> {
+  options: Array<Type>;
+  noItemsText: ReactNode;
+  isLoading?: boolean;
+  optionsListClassName?: string;
+  onOptionChange: (newValue: Type) => void;
+  renderOptionContent: (option: Type) => ReactNode;
+}
+interface SelectOptionsProps<Type> extends SelectOptionsPropsBase<Type> {
+  opened: boolean;
+  setOpened: Dispatch<SetStateAction<boolean>>;
+}
+
+const SelectOptions = <Type extends unknown>({
+  opened,
+  options,
+  noItemsText,
+  isLoading,
+  optionsListClassName,
+  onOptionChange,
+  setOpened,
+  renderOptionContent
+}: SelectOptionsProps<Type>) => {
+  const handleOptionClick = (newValue: Type) => {
+    onOptionChange(newValue);
+    setOpened(false);
+  };
+
+  return (
+    <DropdownWrapper
+      opened={opened}
+      className="origin-top overflow-x-hidden overflow-y-auto"
+      style={{
+        maxHeight: '15.125rem',
+        backgroundColor: 'white',
+        borderColor: '#e2e8f0'
+      }}
+    >
+      {(options.length === 0 || isLoading) && (
+        <div className="my-8 flex flex-col items-center justify-center text-gray-500">
+          {isLoading ? (
+            <Spinner theme="primary" style={{ width: '3rem' }} />
+          ) : (
+            <p className="flex items-center justify-center text-gray-600 text-base font-light">
+              <span>{noItemsText}</span>
+            </p>
+          )}
+        </div>
+      )}
+
+      <ul className={optionsListClassName}>
+        {options.map(option => (
+          <li key={v4()}>
+            <button className="w-full" disabled={(option as any)?.disabled} onClick={() => handleOptionClick(option)}>
```

Comment:
could we avoid assertion here?

Reply:
I have tried but typescript din't recognize generic I wrote. If you know how to do this, let me know